[MenuList] Include disabled items in keyboard navigation

scottander opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- [X] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui-org/material-ui/blob/master/CONTRIBUTING.md#sending-a-pull-request).

Small change to make menulist inline with WCAG specifications. (https://www.w3.org/TR/wai-aria-practices/#kbd_disabled_controls)

Disabled items should not be skipped in the menu lists during keyboard navigation as a visually impaired user will not be correctly notified via screen readers.

eps1lon commented:
I'm inclined to agree here (especially `aria-disabled` should not be checked). But we do need a test for this behavior.

scottander commented:
thanks @eps1lon I have modified the tests to check for the modified behavior.

scottander commented:
> * What visual feedback should we provide when a disabled item is focused? There is no feedback right now.
> * Should the behavior be the default one? In the WAI link, it seems to be referenced as something common, but does it mean it should be the default? For instance, on macOS, disabled items in a menu aren't focusable. The same goes for Google spreadsheets/docs.
> * Should the Autocomplete behave identically? It doesn't right now, it could be inconsistent.

A few responses:

-I don't believe based on the wcag specification that any other visual styling is required for disabled  & focused elements. That would be a library specific choice for MUI.
-I do believe that if the intent of this library is to conform to accessibility guidelines then the WCAG recommended behavior should be the default.
-I think it would make sense that Autocomplete to behave the same way, although I could not find any specific requirement in the w3 spec to confirm or deny this.

Happy to update this PR as necessary though.

scottander commented on packages/material-ui/src/MenuList/MenuList.js:
```diff
@@ -59,12 +59,7 @@ function moveFocus(list, currentFocus, disableListWrap, traversalFunction, textC
-      nextFocus.getAttribute('aria-disabled') === 'true' ||
-      !textCriteriaMatches(nextFocus, textCriteria)
-    ) {
+    if (!nextFocus.hasAttribute('tabindex') || !textCriteriaMatches(nextFocus, textCriteria)) {
```
updated

oliviertassinari commented:
> I do believe that if the intent of this library is to conform to accessibility guidelines then the WCAG recommended behavior should be the default.

@scottander I have a hard time finding UI libraries to benchmark against. Do you have some good sources to benchmark? I only have found weak signals.

The following go against it:
- macOS
- Google Docs
- [reakit](https://reakit.io/docs/menu/#usage)
- [select2](https://select2.org/options#disabling-options)

The following go in favor of it:
- [Kendo jQuery menu bar](https://demos.telerik.com/kendo-ui/menu/keyboard-navigation).
- [react-select](https://react-select.com/home#welcome)

I wonder, should we make a distinction between a "menu bar" and a "dropdown menu"?

scottander commented:
@oliviertassinari I do not have any more good examples of libraries following this pattern, although I will say that when trying to develop products that are accessible and do follow s3 specs closely enough to meet WCAG compliance it is generally not enough to follow what other software is doing, as accessibility is generally an after though (if that). I will add this as a configurable option in the props and add some more tests to cover.

Any preference on the name? "enableFocusForDisabledItems" is what I am currently thinking.

@eps1lon could you weigh in on the default value?

joshwooding commented:
I think the default should be to allow focus on disabled items and follow the guidelines.

scottander commented on packages/material-ui/test/integration/MenuList.test.js:
```diff
@@ -356,71 +356,65 @@ describe('<MenuList> integration', () => {
+    expect(menuitems[2]).to.have.focus;
   });
 
   it('should stay on a single item if it is the only focusable one', () => {
```
with the new prop i can leave this test as it was.

scottander commented:
@joshwooding @oliviertassinari is there anything else I need to do- to move this PR along? Thanks for the help.

diegohaz commented:
I guess the question is: **Is the presence of this element important enough that it should be perceivable even when disabled?**

The answer, obviously, depends on the context. And I guess it can only be answered on the app side. So making it configurable is the best option.

Reakit has a `focusable` prop that can be passed to any `MenuItem` to make them focusable even when they're disabled. On [`Tab`](https://reakit.io/docs/tab/#usage), this prop is `true` by default.

eps1lon commented on docs/pages/api-docs/menu-list.md:
```diff
@@ -31,6 +31,7 @@ the focus is placed inside the component it is fully keyboard accessible.
 | <span class="prop-name">autoFocusItem</span> | <span class="prop-type">bool</span> | <span class="prop-default">false</span> | If `true`, will focus the first menuitem if `variant="menu"` or selected item if `variant="selectedMenu"` |
 | <span class="prop-name">children</span> | <span class="prop-type">node</span> |  | MenuList contents, normally `MenuItem`s. |
 | <span class="prop-name">disableListWrap</span> | <span class="prop-type">bool</span> | <span class="prop-default">false</span> | If `true`, the menu items will not wrap focus. |
+| <span class="prop-name">enableFocusForDisabledItems</span> | <span class="prop-type">bool</span> | <span class="prop-default">true</span> | If `true`, will allow focus on disabled items. |
```
For boolean props we want have `false` as the default value to enable boolean shorthands like `<MenuList disableFocusForDisabledItems />`. See https://material-ui.com/guides/api/#property-naming

scottander replied:
seems like the consensus is to have this turned off by default so I just updated the current prop default.

scottander commented:
> Thanks for pushing the effort further. I think that we still need to explore the following:
> 
> 1. We need visual feedback on the focused disabled items.
> 2. The WAI-ARIA explains the pros and cons very well (even for screen readers only). Now, If we take into account that allowing the focus on disabled items harms the experience of none screen readers users, I think that the pragmatic choice is to optimize for the many, hence keep the current behavior as default. But I would need to benchmark on Windows to get a strong point of view on this. So far, it's only based on my experience on the products I use daily (macOS, Gsuite). [JedWatson/react-select#3902](https://github.com/JedWatson/react-select/issues/3902) is interesting too.
> 3. I think that the behavior should be consistent, hence taking the Autocomplete component into account here.

Thanks @oliviertassinari,
1: I will take a look at this one, is adding a style key for disabled focused items going to be enough?
2: In comments for this component it calls out that it should follow aria best practices, to me it seems clear that the standard indicates that you should make disabled items focusable for options in a list box, tabs, menu items and tree views. I am happy to bend with the wind on this one though. Do we need any more consensus on this or can i move forward on the assumption that this will be an opt-in behavior?
3: Once i have confirmation that this behavior should be opt-in i will updated auto-complete as well to follow suite.

joshwooding commented:
It is also worth thinking that the decision of opt-in vs opt-out affects all the above listed components too.

scottander commented:
@oliviertassinari I believe i have addressed the points you raised, however i can't seem to get the build checks to succeed, it says that i need to run yarn proptypes however this seems to have no effect, any advice or help would be appreciated.

scottander commented:
...bump @oliviertassinari or @joshwooding or @eps1lon

oliviertassinari commented:
@scottander Thanks

scottander commented:
No problem and thanks for the help getting this one over the finish line @oliviertassinari
